Feature/behaviour modularity

funketh opened the pull request.

funketh committed:
make Car more modular by using a separate behaviour class

funketh committed:
use LINQ in DespawnCars()

funketh committed:
initialize behaviour in GameObjectData

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Car.cs:
```diff
@@ -1,33 +1,49 @@
             positionOnRoad += speed;
             var position = road.GetAbsolutePosition(positionOnRoad, lane);
-            carTransform.position = new Vector3(position.x, carTransform.position.y, position.y);
+            gameObject.transform.position = new Vector3(position.x, gameObject.transform.position.y, position.y);
```
Der transform sollte gecached werden

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Vertices/EndPoint.cs:
```diff
@@ -72,35 +72,18 @@ public void SpawnCars()
         public void DespawnCars()
         {
-            foreach(var car in _edge.cars.ToList()) 
+            foreach (var car in _edge.cars.ToList().Where(car => car.positionOnRoad > _edge.length))
```
`>=` ist besser, man entfernt damit auch Autos, die sich genau am Ende befinden

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Car.cs:
```diff
@@ -1,33 +1,49 @@
+
+    public class CarBehaviour : LinkedBehaviour<Car>
+    {
+        private void FixedUpdate()
+        {
+            _data.Move();
+        }
```
Ich dachte, wir hätten uns darauf geeinigt, dass die Move-Funktion zentral aufgerufen wird, um zu gewährleisten, dass alle Autos sich zeitgleich bewegen. Siehe Diskussion in #7

funketh replied:
naja wir werden ja nun sphären nutzen, um die moves zu planen, wobei die kollisionen ja automatisch im physicsupdate gecheckt werden, welches vor dem fixedupdate geschieht

RichardKarlFuchs replied:
Wie hier zu sehen ist, ist das interne physicsupdate nach fixedUpdate: https://docs.unity3d.com/Manual/ExecutionOrder.html

funketh replied:
...was außer für das erste update aufs gleiche hinausläuft.

RichardKarlFuchs replied:
Touché

funketh committed:
use game object id as name so we don't have to keep track of ids manually

RichardKarlFuchs commented on Assets/Scripts/DataTypes/GameObjectData.cs:
```diff
@@ -0,0 +1,25 @@
+
+        protected void CreateGameObject(GameObject prefab, Vector3 position, Quaternion rotation)
+        {
+            gameObject = UnityEngine.Object.Instantiate(prefab, position, rotation);
```
Namespace `UnityEngine` wird bereits bereits verwendet
```suggestion
            gameObject = Object.Instantiate(prefab, position, rotation);
```

funketh replied:
System hat auch ein Object

RichardKarlFuchs commented on Assets/Scripts/DataTypes/GameObjectData.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public void Dispose()
+        {
+            UnityEngine.Object.Destroy(gameObject);
```
Siehe vorherigen Comment
```suggestion
            Object.Destroy(gameObject);
```

funketh committed:
move RoadSpawner functionality into Edge constructor

funketh committed:
cache the Transform objects

funketh committed:
change > to >=

(https://github.com/rrickfox/traffic-sim/pull/61#discussion_r373859485)

funketh committed:
refactor frequencies into own classs

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Frequencies.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public IEnumerable<int> CurrentActiveIndices()
+        {
+            var oldTicks = _ticks++;
+            return _frequencies.Where(frequency => oldTicks % frequency == 0).Select((frequency, index) => index);
```
Wozu muss oldTicks zwischengespeichert werden?

funketh replied:
wegen dem inkrementieren

RichardKarlFuchs replied:
Es wird doch aber nach dem Vergleichen inkrementiert...

funketh replied:
Soll es doch auch?

RichardKarlFuchs replied:
Dann kann es doch direkt in der Where Abfrage aufgerufen werden, oder irre ich mich?

funketh replied:
Das in der Where Abfrage ist ne anonyme Lambda Funktion, die nen bool zurückliefert.
D.h. dass sie wie ne normale Funktion für jede frequency einmal aufgerufen wird.
Würde man dort also ticks inkrementieren, so würde dies für jede frequency 1 mal passieren.

RichardKarlFuchs replied:
Jetzt hat's klick gemacht

funketh committed:
remove erroneous car movement in SimulationController

funketh committed:
modularize Vertex behaviour

- StartPathfinding() had to be moved

funketh committed:
move all pathfinding related functionality into its own namespace

funketh committed:
Merge branch 'develop' into feature/behaviour-modularity

funketh committed:
always create empty game objects, even if they're not doing anything

funketh committed:
give every GameObject a unique default name

funketh committed:
restrict VertexBehaviour to only take IVertex classes

funketh committed:
iterate Vertex.edges instead of the parameter edges

funketh committed:
differentiate between visible and invisible Vertices

- game object creation is now done at construction time
- absolute car position is now calculated inside Car, not Edge

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Car.cs:
```diff
@@ -1,33 +1,58 @@
             this.lane = lane;
+
+            var position = GetAbsolutePosition();
+            transform.position = new Vector3(position.x, prefab.transform.localScale.y / 2 + prefab.transform.localScale.y / 2, position.y);
```
Y-Komponente der Position erzeugt ist falsch, da das Road-Prefab nicht mehr übergeben wird.
Vorschlag: Eine Neue Konstante `ROAD_HEIGHT`, dann lässt sich die y-position wieder berechnen.
```suggestion
            transform.position = new Vector3(position.x, CONSTANTS.ROAD_HEIGHT + prefab.transform.localScale.y / 2, position.y);
```
Dabei ist `CONSTANTS.ROAD_HEIGHT = roadPrefab.localScale.y = 0.05`
Entsprechend muss in `Edge.cs` folgendes gesetzt werden:
```
transform.position.y = CONSTANTS.ROAD_HEIGHT / 2;
transform.localScale.y = CONSTANTS.ROAD_HEIGHT;
```

RichardKarlFuchs committed:
add road height constant, fix calculation of y-position and scale of cars and roads

RichardKarlFuchs committed:
use cached transform

funketh committed:
Merge branch 'develop' into feature/behaviour-modularity
